engine/engine.go

```diff
@@ -50,82 +64,134 @@ func New(ctx context.Context, cliCtx *cli.Context) (*Engine, error) {
 		host:       uuid.New().String()[:8], // TODO: set from config toml
 	}
 
+	// Create context to cancel workers on Close
+	ctx, e.cancel = context.WithCancel(ctx)
+
+	e.wg.Add(workers)
 	for i := 0; i < workers; i++ {
-		go e.worker(i)
+		go e.worker(ctx, i)
 	}
 
 	return e, nil
 }
 
 func (e *Engine) Close() {
+	e.cancel()  // stop workers
+	e.wg.Wait() // wait for workers to stop
 	e.closer()
 }
 
-func (e *Engine) worker(n int) {
-	log.Infow("engine worker started", "worker_id", n)
-
+func (e *Engine) popTask(ctx context.Context) tasks.Task {
+	// Pop tasks until a task is found or until no more tasks
 	for {
```

Comment:
why is this in a loop? the popTask caller already does a loop with 5s waits. if you remove the loop, then the ctx.Err check should also go.

Reply:
This is in a loop to skip over the unusable tasks.  Probably not really necessary, so removed.